File: packages/engine/src/llm-service.ts
// LLM Service with robust fallback and optional dependency loading
export class LLMService {
    private model: any = null;
    private langchainReady: boolean = false;

    constructor() {
        this.initializeLangChain();
    }

    private async initializeLangChain() {
        const apiKey = process.env.OPENAI_API_KEY;
        if (!apiKey) return;

        try {
            // Dynamic check for langchain availability to prevent build/runtime crashes
            const { ChatOpenAI } = require("@langchain/openai");
            this.model = new ChatOpenAI({
                openAIApiKey: apiKey,
                modelName: "gpt-4-turbo-preview",
                temperature: 0.7,
            });
            this.langchainReady = true;
        } catch (error) {
            console.warn("LangChain modules not found or could not be initialized. Using advanced PMP fallback.");
        }
    }

    async generateContent(promptName: string, variables: Record<string, string>, template: string): Promise<string> {
        if (this.langchainReady && this.model) {
            try {
                const { PromptTemplate } = require("@langchain/core/prompts");
                const { StringOutputParser } = require("@langchain/core/output_parsers");
                const { RunnableSequence } = require("@langchain/core/runnables");

                const prompt = PromptTemplate.fromTemplate(template);
                const chain = RunnableSequence.from([
                    prompt,
                    this.model,
                    new StringOutputParser(),
                ]);

                return await chain.invoke(variables);
            } catch (error) {
                console.warn(`LLM Generation failed for ${promptName}, falling back:`, error);
                return this.fallbackGeneration(promptName, variables);
            }
        } else {
            return this.fallbackGeneration(promptName, variables);
        }
    }

    private fallbackGeneration(promptName: string, variables: Record<string, string>): string {
        let { idea, canonical_json } = variables;
        let jsonData: any = null;

        if (canonical_json) {
            try {
                jsonData = JSON.parse(canonical_json);
            } catch (e) {
                console.warn("Fallback: Failed to parse canonical_json");
            }
        }

        if (promptName === 'CANONICAL_JSON') {
            return JSON.stringify({
                project_overview: {
                    summary: idea || "A new digital solution",
                    problem_statement: "Manual processes are inefficient",
                    objectives: ["Automate workflows", "Improve user experience", "Provide data insights"]
                },
                users: {
                    target_users: ["Standard Users", "System Administrators"],
                    user_roles: ["User", "Admin"]
                },
                scope: {
                    in_scope: ["Web interface", "Core database", "User authentication"],
                    out_of_scope: ["Mobile application", "Offline mode", "Legacy system migration"]
                },
                features: {
                    must_have: ["User Dashboard", "Data Entry Forms", "Reporting Module"],
                    nice_to_have: ["Dark Mode", "Push Notifications", "Advanced Analytics"]
                },
                constraints: {
                    time: "12 weeks",
                    budget: "Enterprise standard",
                    technical: "Modern web stack",
                    regulatory: "GDPR Compliance"
                },
                assumptions: ["Internet connectivity is available", "Users have basic technical proficiency"],
                risks: ["Integration challenges", "Data security vulnerabilities"],
                success_metrics: ["90% user adoption", "Reduced processing time"],
                scalability_expectations: "Support up to 10,000 concurrent users"
            }, null, 2);
        }

        const summary = jsonData?.project_overview?.summary || idea || 'Project Alpha';
        const features = jsonData?.features?.must_have || ['User Management', 'Dashboard', 'Data Analytics'];
        const roles = jsonData?.users?.user_roles || ['Admin', 'User'];
        const objectives = jsonData?.project_overview?.objectives || ['Deliver high value', 'Ensure security'];

        if (promptName === 'REQUIREMENTS') {
            return `# Requirement Specification: ${summary}
            
## Purpose
To define the functional and non-functional requirements for the ${summary} system.

## Scope
Includes development of ${features.join(', ')}.

## User Roles
${roles.map((r: string) => `- **${r}**: Primary user role.`).join('\n')}

## Functional Requirements
${features.map((f: string, i: number) => `### FR-0${i + 1}: ${f}
The system shall provide ${f} functionality.`).join('\n\n')}

## Non-Functional Requirements
- **Performance**: System responds within 200ms.
- **Security**: OAuth2 based authentication.

## Assumptions
- Stable network environment.

## Constraints
- Completion within 12 weeks.

## Dependencies
- Cloud provider availability.
`;
        }

        if (promptName === 'PLANNING') {
            return `# Project Planning: ${summary}

## Project Goals
${objectives.map((o: string) => `- ${o}`).join('\n')}

## Deliverables
- Functional Prototype
- Production-Ready Codebase
- Technical Documentation

## Project Phases
1. Initial Analysis
2. Core Development
3. Quality Assurance
4. Launch

## Milestones
- M1: Scope Definition (Week 2)
- M2: MVP Completion (Week 8)
- M3: Final Release (Week 12)

## Resource Overview
- 2 Full-stack Developers
- 1 UI/UX Designer
- 1 QA Engineer

## Communication Plan
Weekly stakeholder updates via email and bi-weekly demo calls.
`;
        }

        if (promptName === 'ARCHITECTURE') {
            return `# Technical Architecture: ${summary}

## System Overview
A robust, scalable web architecture designed for ${summary}.

## High-Level Architecture
Layered architecture: Presentation, Logic, and Data.

## Technology Stack
- **Frontend**: React / Next.js
- **Backend**: Node.js / Express
- **Database**: PostgreSQL
- **Infrastructure**: AWS / Vercel

## Data Flow Overview
Users interact with Frontend -> API calls to Backend -> Transactional data in Database.

## Security Considerations
- TLS 1.3 Encryption
- JWT Session Management
- Role-based Access Control

## Scalability Strategy
Horizontal scaling using containerized microservices.

## Deployment Overview
Automated CI/CD pipeline with Blue/Green deployment strategy.
`;
        }

        if (promptName === 'IPMP') {
            return `# Integrated Project Management Plan (IPMP): ${summary}

## Project Overview
Consolidated execution strategy for the ${summary} project.

## Governance Structure
Project Sponsor -> Project Manager -> Technical Team.

## Scope Management
Change requests evaluated by CCB (Change Control Board).

## Schedule Management
12-week timeline with bi-weekly sprints.

## Cost Management
Managed within the enterprise operational budget.

## Quality Management
Unit tests, integration tests, and UAT.

## Risk Management
Continuous monitoring and mitigation of technical risks.

## Change Control
Version-controlled codebase and documented requirement changes.

## Communication Management
Centralized project dashboard for all stakeholders.
`;
        }

        if (promptName === 'SCHEDULE_COST') {
            return `# Schedule & Cost Estimation: ${summary}

## Work Breakdown Structure (High-Level)
1. Environment Setup
2. Module Development (${features.slice(0, 2).join(', ')})
3. Integration
4. Testing

## Estimated Timeline
12 weeks total execution time.

## Resource Effort Estimates
- Dev: 480 hours
- QA: 160 hours
- PM: 80 hours

## Cost Breakdown
- Labor: $60,000
- Infrastructure: $5,000
- Licenses: $2,000

## Contingency Planning
15% budget buffer for unforeseen technical challenges.
`;
        }

        if (promptName === 'QUALITY_RISK') {
            return `# Quality & Risk Management: ${summary}

## Quality Objectives
Zero critical bugs at launch; 99.9% uptime.

## Quality Standards
ISO/IEC 25010 Software Quality Standards.

## Review & Approval Process
Peer code reviews and automated linting.

## Risk Register
- RISK-01: Delay in 3rd party API (Medium)
- RISK-02: Data breach (Low probability, High impact)

## Risk Mitigation Plan
Early prototyping and rigorous security audits.
`;
        }

        if (promptName === 'TESTING_RELEASE') {
            return `# Testing Strategy: ${summary}

## Test Strategy
Risk-based testing approaching focused on critical paths.

## Test Types
- Unit Testing
- Integration Testing
- System Integration (SIT)
- User Acceptance (UAT)

## Test Scenarios (High-Level)
- Successful user login
- ${features[0]} execution
- Error handling on invalid input

## Entry & Exit Criteria
- Entry: 100% code completion
- Exit: 0 critical defects remaining

## Defect Management
Jira-based tracking with prioritization matrix.

## Release Approval
Final sign-off by Product Owner and QA Lead.
`;
        }

        if (promptName === 'UI_UX') {
            const screens = ['Dashboard', 'Detail View', 'Settings', 'Profile', 'Activity Log'];
            return `# UI/UX Design Specification: ${summary}

${screens.map(s => `### ${s} Screen
- **Purpose**: Key interface for ${s}.
- **Roles**: ${roles.join(', ')}
- **Components**: Navigation, Header, Content Area
- **Interactions**: Button clicks, Form submission
- **States**: Loading, Active, Error`).join('\n\n')}
`;
        }

        return `# ${promptName} Document for ${summary}
        
Professional documentation generated for project ${summary}.
`;
    }
}

File: apps/api-py/engine/llm_service.py
import os
import json
from datetime import datetime
from typing import Dict, Any, Optional

class LLMService:
    VERSION = "1.1.0-PY"

    def __init__(self):
        self.api_key = os.getenv("OPENAI_API_KEY")
        self.client = None
        if self.api_key:
            try:
                from openai import OpenAI
                self.client = OpenAI(api_key=self.api_key)
            except ImportError:
                print("OpenAI module not found. Using fallback.")

    async def generate_content(self, prompt_name: str, variables: Dict[str, str], template: str) -> str:
        if self.api_key and self.client:
            try:
                # Format template with variables
                formatted_prompt = template
                for k, v in variables.items():
                    formatted_prompt = formatted_prompt.replace(f"{{{k}}}", str(v))
                
                response = self.client.chat.completions.create(
                    model="gpt-4-turbo-preview",
                    messages=[{"role": "user", "content": formatted_prompt}],
                    temperature=0.7
                )
                return response.choices[0].message.content
            except Exception as e:
                print(f"LLM Generation failed for {prompt_name}, falling back: {e}")
                return self.fallback_generation(prompt_name, variables)
        else:
            return self.fallback_generation(prompt_name, variables)

    def fallback_generation(self, prompt_name: str, variables: Dict[str, str]) -> str:
        canonical_json = variables.get('canonical_json')
        idea = variables.get('idea', 'Project Idea')
        
        json_data = None
        if canonical_json:
            try:
                json_data = json.loads(canonical_json)
            except:
                pass

        if prompt_name == 'CANONICAL_JSON':
            return json.dumps({
                "project_overview": {
                    "summary": idea,
                    "problem_statement": "Manual processes in " + idea + " are inefficient.",
                    "objectives": ["Automate workflows", "Improve user experience", "Provide data insights"]
                },
                "users": {
                    "target_users": ["Internal Staff", "System Administrators"],
                    "user_roles": ["Standard User", "Admin"]
                },
                "scope": {
                    "in_scope": ["Web interface", "Core database", "User authentication"],
                    "out_of_scope": ["Mobile application", "Offline mode"]
                },
                "features": {
                    "must_have": ["User Dashboard", "Data Entry Forms", "Reporting Module"],
                    "nice_to_have": ["Dark Mode", "Push Notifications"]
                },
                "constraints": {
                    "time": "12 weeks",
                    "budget": "Enterprise standard",
                    "technical": "Modern web stack",
                    "regulatory": "GDPR Compliance"
                },
                "assumptions": ["Stable internet connectivity", "Basic technical proficiency"],
                "risks": ["Data security vulnerabilities", "Integration challenges"],
                "success_metrics": ["90% user adoption", "Reduced processing time"],
                "scalability_expectations": "Support up to 10,000 users"
            }, indent=2)

        # Extraction from JSON for other docs
        summary = idea
        objectives = ["High quality delivery", "User satisfaction"]
        features = ["Core System"]
        roles = ["Admin", "User"]

        if json_data:
            summary = json_data.get("project_overview", {}).get("summary", summary)
            objectives = json_data.get("project_overview", {}).get("objectives", objectives)
            features = json_data.get("features", {}).get("must_have", features)
            roles = json_data.get("users", {}).get("user_roles", roles)

        if prompt_name == 'REQUIREMENTS':
            return f"""# Requirement Document: {summary}
            
## 1. Project Background & Objectives
{summary}
Objectives: {', '.join(objectives)}

## 2. Functional Requirements
{chr(10).join([f'- REQ-{i+1:03}: {f}' for i, f in enumerate(features)])}

## 3. User Roles
{', '.join(roles)}
"""

        if prompt_name == 'PLANNING':
            return f"""# Project Planning Document: {summary}
            
## 1. Work Breakdown Structure
- Phase 1: Requirement & Design
- Phase 2: Core Development of {features[0] if features else 'system'}
- Phase 3: Integration & Testing

## 2. Milestones
- M1: Sign-off (Week 2)
- M2: MVP (Week 8)
- M3: Launch (Week 12)
"""

        if prompt_name == 'ARCHITECTURE':
            return f"""# Technical Architecture: {summary}
            
## 1. Tech Stack
- Frontend: Next.js
- Backend: Python (FastAPI)
- Database: PostgreSQL

## 2. System Components
- API Server
- Auth Service
- Data Storage Service
"""

        if prompt_name == 'IPMP':
            return f"""# Integrated Project Management Plan (IPMP): {summary}
            
## 1. Purpose
Defines the integration of all project work for {summary}.

## 2. Success Criteria
{chr(10).join([f'- {o}' for o in objectives])}
"""

        if prompt_name == 'SCHEDULE_COST':
            return f"""# Schedule & Cost Plan: {summary}
            
## 1. Timeline
Total Project Duration: 12 weeks.

## 2. Budget Allocation
- Development: 70%
- Design: 15%
- Infrastructure: 15%
"""

        if prompt_name == 'QUALITY_RISK':
            return f"""# Quality, Risk & Procurement Plan: {summary}
            
## 1. Quality Standards
Adherence to clean code principles and ISO quality metrics.

## 2. Risk Matrix
- R1: Scope creep (Medium)
- R2: Tech debt (Low)
"""

        if prompt_name == 'TESTING_RELEASE':
            return f"""# Testing & Release Plan: {summary}
            
## 1. Testing Strategy
- Unit Testing
- Integration Testing
- User Acceptance Testing (UAT)

## 2. Deployment Plan
CI/CD based automated deployment to production.
"""

        if prompt_name == 'UI_UX':
            return f"""# UI/UX Design Specification: {summary}
            
## 1. Screen List
- Dashboard for {roles[0]}
- Data Entry for {roles[1] if len(roles) > 1 else 'Users'}
- Global Search

## 2. Style Guide
Modern, clean interface with blue-primary color palette.
"""

        return f"# {prompt_name} Document for {summary}\n\nDetailed specifications generated for the project."
